chains/ethereum/forwarder_client.go

```diff
@@ -0,0 +1,219 @@
+// If transaction usage of a nonce failes the nonce shouldnt be
+// updated upon unlock, instead nil should be supplied
+func (c *ForwarderClient) UnlockAndSetNonce(nonce *big.Int) {
+	if nonce != nil {
+		c.forwarderNonce = nonce
+	}
+	c.forwarderNonceLock.Unlock()
+}
+
+// Generate the 712 type hash for signing
+func (c *ForwarderClient) TypedHash(
+	from, to string,
+	data []byte,
+	value, gas *math.HexOrDecimal256,
+	nonce *big.Int,
+	chainId *math.HexOrDecimal256,
+	verifyingContract string) ([]byte, error) {
+
+	typedData := signer.TypedData{
+		Types: signer.Types{
+			// EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"
+			"EIP712Domain": []signer.Type{
+				{Name: "name", Type: "string"},
+				{Name: "version", Type: "string"},
+				{Name: "chainId", Type: "uint256"},
+				{Name: "verifyingContract", Type: "address"},
+			},
+			// ForwardRequest(address from,address to,uint256 value,uint256 gas,uint256 nonce,bytes data)
+			"ForwardRequest": []signer.Type{
+				{Name: "from", Type: "address"},
+				{Name: "to", Type: "address"},
+				{Name: "value", Type: "uint256"},
+				{Name: "gas", Type: "uint256"},
+				{Name: "nonce", Type: "uint256"},
+				{Name: "data", Type: "bytes"},
+			},
+		},
+		PrimaryType: "ForwardRequest",
+		Domain: signer.TypedDataDomain{
+			Name:              "MinimalForwarder",
+			ChainId:           chainId,
+			Version:           "0.0.1",
+			VerifyingContract: verifyingContract,
+		},
+		Message: signer.TypedDataMessage{
+			"from":  from,
+			"to":    to,
+			"value": value,
+			"gas":   gas,
+			"data":  data,
+			"nonce": math.NewHexOrDecimal256(nonce.Int64()),
+		},
+	}
+
+	domainSeparator, err := typedData.HashStruct("EIP712Domain", typedData.Domain.Map())
+	if err != nil {
+		return nil, err
+	}
+	typedDataHash, err := typedData.HashStruct(typedData.PrimaryType, typedData.Message)
+	if err != nil {
+		return nil, err
+	}
+	rawData := []byte(fmt.Sprintf("\x19\x01%s%s", string(domainSeparator), string(typedDataHash)))
+	return crypto.Keccak256(rawData), nil
+}
+
+// Pack and sign forwarder contract 'execute' arguments
+func (c *ForwarderClient) PackAndSignForwarderArg(
+	from, to common.Address,
+	data []byte,
+	nonce, value, gas *big.Int,
+	chainId uint,
+	kp secp256k1.Keypair) ([]byte, error) {
+	forwarderHash, err := c.TypedHash(
+		from.String(),
+		to.String(),
+		data,
+		math.NewHexOrDecimal256(int64(value.Uint64())),
+		math.NewHexOrDecimal256(int64(gas.Uint64())),
+		nonce,
+		math.NewHexOrDecimal256(int64(chainId)),
+		c.forwarderAddress.String())
+
+	if err != nil {
+		return nil, err
+	}
+
+	// now sign that hash
+	sig, err := crypto.Sign(forwarderHash, kp.PrivateKey())
+	sig[64] += 27 // Transform V from 0/1 to 27/28
+
+	if err != nil {
+		return nil, err
+	}
+
+	// struct ForwardRequest {
+	//     address from;
+	//     address to;
+	//     uint256 value;
+	//     uint256 gas;
+	//     uint256 nonce;
+	//     bytes data;
+	// }
```

Comment:
Why not declare this type instead of creating it anonymously below? Declaring the type would be self-documenting. I believe you can declare types within functions as well.